src/Agda2Hs/Compile/Term.hs

```diff
@@ -20,313 +21,448 @@ import Agda.Syntax.Internal
-isSpecialTerm q = case prettyShow q of
-  _ | isExtendedLambdaName q                    -> Just lambdaCase
-  "Haskell.Prim.if_then_else_"                  -> Just ifThenElse
-  "Haskell.Prim.the"                            -> Just expTypeSig
+
+-- * Compilation of special definitions
+
+type DefCompileRule = Type -> [Term] -> C (Hs.Exp ())
+
+isSpecialDef :: QName -> Maybe DefCompileRule
+isSpecialDef q = case prettyShow q of
+  _ | isExtendedLambdaName q     -> Just (lambdaCase q)
+  "Haskell.Prim.if_then_else_"   -> Just ifThenElse
+  "Haskell.Prim.case_of_"        -> Just caseOf
+  "Haskell.Prim.the"             -> Just expTypeSig
+  "Haskell.Extra.Delay.runDelay" -> Just compileErasedApp
+  _                              -> Nothing
+
+
+-- | Compile a @\where@ to the equivalent @\case@ expression.
+lambdaCase :: QName -> DefCompileRule
+lambdaCase q ty args = setCurrentRangeQ q $ do
+  Function
+    { funClauses = cls
+    , funExtLam  = Just ExtLamInfo {extLamModule = mname}
+    } <- theDef <$> getConstInfo q
+
+  npars <- size <$> lookupSection mname
+
+  let (pars, rest) = splitAt npars args
+      cs           = applys cls pars
+
+  ty' <- piApplyM ty pars
+
+  cs <- mapMaybeM (compileClause' (qnameModule q) (hsName "(lambdaCase)") ty') cs
+
+  case cs of
+    -- If there is a single clause and all patterns got erased, we
+    -- simply return the body.
+    [Hs.Match _ _ [] (Hs.UnGuardedRhs _ rhs) _] -> return rhs
+    _ -> do
+      lcase <- hsLCase =<< mapM clauseToAlt cs -- Pattern lambdas cannot have where blocks
+      eApp lcase <$> compileArgs ty' rest
+      -- undefined -- compileApp lcase (undefined, undefined, rest)
+
+
+-- | Compile @if_then_else_@ to a Haskell @if ... then ... else ... @ expression.
+ifThenElse :: DefCompileRule
+ifThenElse ty args = compileArgs ty args >>= \case
+  -- fully applied
+  b : t : f : es' -> return $ Hs.If () b t f `eApp` es'
+  -- partially applied
+  _               -> genericError "if_then_else_ must be fully applied"
+
+
+-- | Compile @case_of_@ to Haskell @\case@ expression.
+caseOf :: DefCompileRule
+caseOf ty args = compileArgs ty args >>= \case
+  -- applied to pattern lambda (that we remove, hence decrementLCase)
+  e : Hs.LCase _ alts : es' -> decrementLCase $> eApp (Hs.Case () e alts) es'
+  -- applied to regular lambda
+  e : Hs.Lambda _ (p : ps) b : es' ->
+    let lam [] = id
+        lam qs = Hs.Lambda () qs
+    in return $ eApp (Hs.Case () e [Hs.Alt () p (Hs.UnGuardedRhs () $ lam ps b) Nothing]) es'
+  _ -> genericError "case_of_ must be fully applied to a lambda term."
+
+
+-- | Compile @the@ to an explicitly-annotated Haskell expression.
+expTypeSig :: DefCompileRule
+expTypeSig ty args@(_:typ:_:_) = do
+    annot    <- compileType typ
+    exp:args <- compileArgs ty args
+    pure (Hs.ExpTypeSig () exp annot `eApp` args)
+expTypeSig _ _ = genericError "`the` must be fully applied"
+
+
+-- should really be named compileVar, TODO: rename compileVar
```

Comment:
I suggest `s/compileVar/compileVarName` and then `s/compileV/compileVar`.

Reply:
`compileVar` as renamed to `compileDBVar` in #277 . Renamed `compileV` to `compileVar`, lmk if that's enough.